src/askui_runner/modules/core/infrastructure/files/askui.py

```diff
@@ -41,13 +57,158 @@ def __init__(self, base_url: str, headers: dict[str, str]):
+                for pattern in self.HIDDEN_FILES_PATTERNS
+            ):
+                continue
+            remote_files[content.path[len(remote_dir_path) :].lstrip("/")] = content
+
+        # List local files
+        local_files = self._list_local_files(local_dir_path)
+
+        # Create lookup table
+        all_paths = set(local_files.keys()) | set(remote_files.keys())
+
+        # Action mapper
+        action_mapper: Dict[str, Callable[..., None]] = {
+            "upload": lambda local_file_path, remote_file_path: (
+                logging.info(f"Dry: Upload {local_file_path} to cloud")
+                if dry
+                else self._upload_file(local_file_path, remote_file_path)
+            ),
+            "download": lambda remote_file_url, remote_file_last_modified, local_path: (
+                logging.info(f"Dry: Download file to {local_path}")
+                if dry
+                else self._download_file(
+                    remote_file_url, local_path, remote_file_last_modified
+                )
+            ),
+            "delete_local": lambda local_file_path: (
+                logging.info(f"Dry: Delete local file {local_file_path}")
+                if dry
+                else os.remove(local_file_path)
+            ),
+            "delete_remote": lambda remote_file_path: (
+                logging.info(f"Dry: Delete remote file {remote_file_path}")
+                if dry
+                else self._delete_file(remote_file_path)
+            ),
+            "skip": lambda relative_path: logging.info(
+                f"Skip {relative_path} (no changes)"
+            ),
+        }
+
+        for relative_path in sorted(all_paths):
+            local_file: Union[LocalFileDto | None] = local_files.get(relative_path)
+            remote_file: Union[FileDto | None] = remote_files.get(relative_path)
+            remote_file_path = f"{remote_dir_path}/{relative_path}".replace("\\", "/")
+
+            # File exists in both locations
+            if local_file and remote_file:
+                local_mtime = local_file.last_modified
+                remote_mtime = (
+                    datetime.strptime(remote_file.last_modified, "%Y-%m-%dT%H:%M:%SZ")
+                    .replace(tzinfo=timezone.utc)
+                    .timestamp()
+                )
+
+                if source_of_truth == "local" and (
+                    local_mtime > remote_mtime or local_file.size != remote_file.size
+                ):
+                    action_mapper["upload"](local_file.path, remote_file_path)
+
+                elif source_of_truth == "remote" and (
+                    remote_mtime > local_mtime or local_file.size != remote_file.size
+                ):
+                    local_path = os.path.join(local_dir_path, relative_path)
+                    action_mapper["download"](
+                        remote_file.url, remote_file.last_modified, local_path
+                    )
+                else:
```

Comment:
Just out of interest: What do you think about also comparing the ETags (I could add them to the api)? (see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/list_objects_v2.html)

Reply:
Decided against it for now, maybe future iteration